libraries/azqueue/test/slow_test.js

```diff
@@ -0,0 +1,177 @@
+const assert = require('assert').strict;
+
+helper.dbSuite(path.basename(__filename), function() {
+  const schema = Schema.fromDbDirectory(path.join(__dirname, 'db'));
+  const serviceName = 'test-azqueue';
+  helper.withDb({ schema, serviceName, clearBeforeTests: false });
+
+  suiteSetup(function() {
+    // HINT: run these tests with the following to see explains in the logs:
+    //
+    // docker run -ti -p 127.0.0.1:5432:5432  --rm postgres:11 \
+    //   -c log_statement=all \
+    //   -c session_preload_libraries=auto_explain \
+    //   -c auto_explain.log_min_duration=0s \
+    //   -c auto_explain.log_nested_statements=on \
+    //   -c auto_explain.log_analyze=on
+    //
+    // See below for important things to look for in the explain output.
+    //
+    // These tests will almost certainly pass; don't bother running them
+    // unless you are looking at the explain output.
+
+    // Comment this line out to run these tests.
+    this.skip();
+  });
+
+  const clearDb = async client => {
+    await client.query('delete from azure_queue_messages');
+  };
+
+  const fillDb = async (client, {queues, count, visible, inserted}) => {
+    await client.query(`create temporary table queues (queue_name text)`);
+    for (let queue of queues) {
+      await client.query(`insert into queues values ($1)`, [queue]);
+    }
+
+    if (!inserted) {
+      inserted = new Date();
+    }
+
+    if (!visible) {
+      visible = taskcluster.fromNow('-1 seconds');
+    }
+
+    await client.query(`
+      insert into azure_queue_messages (
+          queue_name,
+          message_id,
+          message_text,
+          inserted,
+          visible,
+          expires
+        ) select
+            queue_name,
+            public.gen_random_uuid() as message_id,
+            'msg-' || series.i::text as message_text,
+            $1 as inserted,
+            $2 as visible,
+            now () + interval '1 hour' as expires
+          from generate_series(0, $3) as series(i), queues
+        `, [inserted, visible, count]);
+    await client.query(`drop table queues`);
+  };
+
+  const analyze = async client => {
+    await client.query('analyze');
+  };
+
+  // For all of these, we should see an index scan using the .._inserted
+  // index.  That index does two things: selects the specific queue (avoiding
+  // one huge queue slowing down other queries), and gets messages in the
+  // correct FIFO order (avoiding a sort).  We still scan over those messages
+  // sequentally to skip any that are invisible or expired.
+  //
+  // Room for improvement: this currently scans rows looking at visibility
+  // and expiration; ideally postgres could do that with an index.  The
+  // planner does not do so automatically, probably because of the LIMIT: combining
+  // two indexes would require scanning both and creating a bitmap, when in many
+  // cases only a partial index scan will be required to satisfy the LIMIT.
```

Comment:
Does this answer my question from above about indices?

Reply:
Can we rely on the planner always Doing the Right Thing :tm: ?

Reply:
> Does this answer my question from above about indices?

indirectly

> Can we rely on the planner always Doing the Right Thing tm ?

Not necessarily, but it usually has reasons for doing what it does.  Sometimes those reasons are wrong (bad heuristics).